src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py

```diff
@@ -80,113 +124,301 @@ def predict(self, document: Document) -> List[SpanGroup]:
         "Please provide custom models."
         """
         self._doc_field_checker(document)
-        words = []
-        local_dictionary = self._build_document_local_dictionary(document)
 
-        skip_first_token = False
-
-        for curr_row, next_row in self._row_pairs(document=document):
-
-            # Skip first token and there is only one token -> skip entire row
-            if skip_first_token and len(curr_row.tokens) == 1:
-                skip_first_token = False
-                continue
-
-            # Add all tokens except last to document words
-            for i, token in enumerate(curr_row.tokens[:-1]):
-                if skip_first_token and i == 0:
-                    skip_first_token = False
-                    continue
+        # TODO: precompute links between start & end token_ids
+        # 1) precompute features
+        token_id_to_ws_token_id, ws_token_id_to_token_ids, token_id_to_token_ids, \
+        row_start_after_hyphen_token_ids, row_end_with_hyphen_token_ids, \
+        max_row_end_token_id_to_min_row_start_token_id, punct_strip_candidate_token_ids = \
+            self._precompute_token_features(document=document)
+
+        # 2) build dictionary
+        internal_dictionary = self._build_internal_dictionary(
+            document=document,
+            token_id_to_token_ids=token_id_to_token_ids,
+            row_start_after_hyphen_token_ids=row_start_after_hyphen_token_ids,
+            row_end_with_hyphen_token_ids=row_end_with_hyphen_token_ids
+        )
+
+        # 3) predict words for using token features
+        token_id_to_word_id, word_id_to_text = self._predict_tokens(
+            document=document,
+            internal_dictionary=internal_dictionary,
+            token_id_to_token_ids=token_id_to_token_ids,
+            row_start_after_hyphen_token_ids=row_start_after_hyphen_token_ids,
+            row_end_with_hyphen_token_ids=row_end_with_hyphen_token_ids,
+            max_row_end_token_id_to_min_row_start_token_id=max_row_end_token_id_to_min_row_start_token_id,
+            punct_strip_candidate_token_ids=punct_strip_candidate_token_ids
+        )
+
+        # 4) transformation
+        words: List[SpanGroup] = self._convert_to_words(
+            document=document,
+            token_id_to_word_id=token_id_to_word_id,
+            word_id_to_text=word_id_to_text
+        )
+
+        # 5) cleanup
+        document.remove(field_name='_ws_tokens')
+        return words
 
-                words.append(self._copy_token_with_text(token))
+    def _precompute_token_features(self, document: Document) -> Tuple:
```

Comment:
Any chance this function can be broken up? It is hard to follow the logic

Reply:
i've split into 2 - whitespace tokenization & token-level features